message/router/middleware/timeout.go

```diff
@@ -9,11 +9,16 @@ import (
 
 // Timeout makes the handler cancel the incoming message's context after a specified time.
 // Any timeout-sensitive functionality of the handler should listen on msg.Context().Done() to know when to fail.
+// Before exiting this middleware, message's original context is restored to not disrupt the logic of possible
+// enveloping handlers (e.g. Retry) that also rely on message context.
 func Timeout(timeout time.Duration) func(message.HandlerFunc) message.HandlerFunc {
 	return func(h message.HandlerFunc) message.HandlerFunc {
 		return func(msg *message.Message) ([]*message.Message, error) {
-			ctx, cancel := context.WithTimeout(msg.Context(), timeout)
+			orgCtx := msg.Context()
+
+			ctx, cancel := context.WithTimeout(orgCtx, timeout)
 			defer func() {
+				msg.SetContext(orgCtx)
```

Comment:
@zanvd how about we use `context.WithoutCancel(ctx)` here? This way the cancellation won't be propagated while the values stored in the context can be. Would that also solve the root issue? 🤔

Reply:
Ah, you're right.

My point is that restoring the context will also remove all context values. If someone uses middleware that uses the context after the handler is executed, it will be lost this way.

Reply:
Here is the way then


```go
orgCtx := msg.Context() 
ctx, cancel := context.WithTimeout(orgCtx, timeout)

defer func() {
   cancel()
   ctx, cancel2 := context.WithCancelWithCause(context.WithoutCancel(ctx))
   msg.SetContext(ctx)
   go func() {
        <- ctx.Done()
        cancel2(ctx.Err)
   }
}()
```

But I find it ugly

Reply:
I'd have to test it out, but this is more or less why I was asking on Discord about usage of context within `Message` - makes it sort of a global object that anyone can modify. Since the blueprint for this was the net/http package, I'd follow its [logic](https://cs.opensource.google/go/go/+/refs/tags/go1.23.3:src/net/http/request.go;l=332) and not even allow changing of the `Message`'s context. Not something to solve in this PR, though.

What would be an example of values stored in the context that we should be vary of? Perhaps copying them to the original context might be a temporary solution.